pkg/scheduler/framework/plugins/examples/stateful/stateful.go

```diff
@@ -47,14 +47,25 @@ func (mp *MultipointExample) Name() string {
 	return Name
 }
 
-// Reserve is the functions invoked by the framework at "reserve" extension point.
+// Reserve is the function invoked by the framework at "reserve" extension point.
 func (mp *MultipointExample) Reserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
 	// Reserve is not called concurrently, and so we don't need to lock.
 	mp.numRuns++
 	return nil
 }
 
-// PreBind is the functions invoked by the framework at "prebind" extension point.
+// Unreserve is the function invoked by the framework when any error happens
+// during "reserve" extension point or later.
+func (mp *MultipointExample) Unreserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) {
+	// Unlike Reserve, the Unreserve method may be called concurrently since
+	// there is no guarantee that there will only one unreserve operations at any
+	// given point in time (for example, during the binding cycle).
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+	mp.numRuns++
```

Comment:
not sure this example make much sense now, and it doesn't emphasize the point that this function should be idempotent.

Reply:
Can you elaborate? I think the example still shows plugin state can be modified at different extension points. Just like before, the Unreserve happens to be so that it may be run concurrently for different pods (just like the Prebind extension point). What has changed in terms of showcasing that feature?

Also, we do not require the Unreserve operation to be idempotent. We will never call the Unreserve function more than once for the same pod, so the authors do not need to account for that, no?

Reply:
> Can you elaborate? I think the example still shows plugin state can be modified at different extension points.

Unreserve is supposed to reverse the effect of Reserve. Both increasing a counter in this example doesn't communicate this expectation.

>Also, we do not require the Unreserve operation to be idempotent. We will never call the Unreserve function more than once for the same pod, so the authors do not need to account for that, no?

We discussed in the past the semantics of Unreserve being idempotent, while it is true that currently we may not be calling this function more than once in the current workflow, I don't know if we can guarantee that in the future. It is easy for plugins to make this function idempotent, so I don't see the point of promising strict one-to-one call with reserve here, especially that this is an error handling extension point.

Reply:
>Unreserve is supposed to reverse the effect of Reserve. Both increasing a counter in this example doesn't communicate this expectation.

Revisiting the full code once more, it looks like there are other things that could be reworkd -- I've changed the example a bit more to make it better.

re idempotency: I've added a comment in the ReservePlugin interface's Unreserve method documentation stating that the implementation must be idempotent.